src/AMAMM.sol

```diff
@@ -2,69 +2,196 @@
+        return 1 hours;
+    }
+
+    function MIN_BID_MULTIPLIER(PoolId) internal view virtual returns (uint256) {
+        return 1.1e18;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    mapping(PoolId id => uint40) internal _lastUpdatedEpoch;
+    mapping(Currency currency => uint256) internal _totalFees;
+    mapping(PoolId id => mapping(uint40 => Bid)) public poolEpochManager;
+    mapping(address manager => mapping(PoolId id => uint256)) internal _refunds;
+    mapping(address manager => mapping(Currency currency => uint256)) internal _fees;
+    mapping(address deposits => uint256) public _userDeposits;
+    mapping(address refunds => uint256) public _userRefunds;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getManager(PoolId id, uint40 epoch) public view returns (Bid memory) {
+        return poolEpochManager[id][epoch];
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint128 deposit, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        // ensure bid is valid
+        // - manager can't be zero address
+        // - bid needs to be greater than the next bid by >10%
+        // - deposit needs to cover the rent for K hours
+        // - deposit needs to be a multiple of rent
+        // - payload needs to be valid
+        if (
+            rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id)) || deposit < rent * K(id)
+                || deposit % rent != 0 || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        // Check if the bid already exists
+        Bid memory newBid = Bid({bidder: msgSender, payload: payload, rent: rent, deposit: deposit});
+
+        if (poolEpochManager[id][_epoch].deposit < deposit) {
+            _userRefunds[poolEpochManager[id][_epoch].bidder] += poolEpochManager[id][_epoch].deposit;
+            poolEpochManager[id][_epoch] = Bid({bidder: msgSender, payload: payload, rent: rent, deposit: deposit});
+            _userDeposits[msgSender] += deposit;
+        }
+
+        _updateEpochBids();
+    }
+
+    /// @inheritdoc IAmAmm
+    function withdrawFromBid(PoolId id, uint40 _epoch, uint128 _amount) external virtual override {
```

Comment:
I don't think we need deposit in the epoch. It would just be one global userDeposits. And each epoch would just have rent, which indicates how much would be charging and how much need to bid over it.